Added Async and FxCop analyzers to Streaming project

gabog opened the pull request:
Relates to #2367 

Completed

- Added CultureInfo and StringComparison to some string operations.
- Added some #pragma disable statements for things we can't change because if source compat.
- Added Dispace call for _dataAvailable in PayloadStream.
- Completed dispose pattern implementation for NamedPipeClient, NamedPipeTransport, WebSocketClient, WebSocketTransport.
- Added Asyn suffix to some private methods for clarity.

TODO (to review and fix with @carlosscastro:

- [x] PayloadReceiver and PayloadSender should probably be IDisposable so they can dispose the instances of ITransportReceiver that they receive (but is not clear what the disconnect logic should do).
- [x] If we make these two objects disposable, we should dispose them in the protected Dispose of NamedPipeClient.
- [x] NamedPipeServer should probably be disposable and dispose the instance of PayloadSender and PayloadReceiver that it creates. 
- [x] WebSocketServer should probably be disposable and dispose the instance of WebSocketTransport that it creates. This will also force us to analyze the usage of WebSocketServer and make the users of this class disposable as well. 
- [x] it is not clear who should dispose the instance of WebSocketTransport that is created in WebSocketClient.ConnectAsyncEx() (the instance is passed to the PayloadSender and the PayloadReceiver. 
- [x] TransportDisconnectedException is internal and being used to control the flow of PayloadReceiver.ReceivePacketsAsync, we need to analyze this and see if we can cleanup the logic so we don't use exception to control execution flow.
- [x] Need to add justification for CA1031 (catch all exceptions)

Here is a screenshot of the issues that need to be fixed:
![image](https://user-images.githubusercontent.com/12264946/88578277-47c25f00-d016-11ea-9a62-9efb9487037a.png)

carlosscastro commented:
@DDEfromOR Could you take a look at this? I did a bunch of the changes so it would be nice to have a third party review this. Thanks!

tomlm commented on libraries/Microsoft.Bot.Streaming/Transport/WebSocket/WebSocketServer.cs:
```diff
@@ -104,8 +106,56 @@ public Task<ReceiveResponse> SendAsync(StreamingRequest request, CancellationTok
         {
-            _sender.Disconnect();
-            _receiver.Disconnect();
+            _sender?.Disconnect();
```
>Disconnect [](start = 21, length = 10)

assign to null to incase someone calls disconnect twice?

tomlm commented on libraries/Microsoft.Bot.Streaming/Utilities/Background.cs:
```diff
@@ -85,20 +89,18 @@ public static void RunForever(Func<CancellationToken, Task<TimeSpan>> task, Time
             catch (Exception)
+#pragma warning restore CA1031 // Do not catch general exception types
             {
             }
```
Why swallow this?

tomlm commented on libraries/Microsoft.Bot.Streaming/Utilities/Background.cs:
```diff
@@ -70,7 +72,9 @@ public static void RunForever(Func<CancellationToken, Task<TimeSpan>> task, Time
                     catch (Exception)
+#pragma warning restore CA1031 // Do not catch general exception types
                     {
                     }
```
why swallow?

DDEfromOR replied:
If we could pipe an  Ilogger in and use properties/eventname the way they were intended, to log failures, we could clean this entire class up a bit. With the IDisposable chain now in place we could actually handle exceptions thrown here and clean up the assemblers/disassemblers that would be derailed by them, too. In either case though we're courting behavior changes that would need more testing than the improvements on this PR, so I'd ask we not block on it for now.

tomlm commented on libraries/Microsoft.Bot.Streaming/Payloads/Disassemblers/PayloadDisassembler.cs:
```diff
@@ -44,7 +44,9 @@ public PayloadDisassembler(IPayloadSender sender, Guid id)
         public async Task DisassembleAsync(CancellationToken cancellationToken = default(CancellationToken))
+#pragma warning restore CA1801 // Review unused parameters
         {
             var w = await GetStreamAsync().ConfigureAwait(false);
```
>GetStreamAsync [](start = 26, length = 14)

should be passing cancelation token to this.

gabog replied:
The abstract method doesn't have cancellationToken, let me see if adding it passes the binary compat check

gabog replied:
Can't add cancellationToken to GetStreamingAsync without breaking binary compat :(

tomlm commented:
>             await SendAsync().ConfigureAwait(false);

same

---
Refers to: libraries/Microsoft.Bot.Streaming/Payloads/Disassemblers/PayloadDisassembler.cs:57 in 625103c. [](commit_id = 625103c45f990b4c8d3a3d281dd47b5ae51461f2, deletion_comment = False)

tomlm commented:
>         private Task SendAsync()

cancelationToken

---
Refers to: libraries/Microsoft.Bot.Streaming/Payloads/Disassemblers/PayloadDisassembler.cs:104 in 625103c. [](commit_id = 625103c45f990b4c8d3a3d281dd47b5ae51461f2, deletion_comment = False)

tomlm commented on libraries/Microsoft.Bot.Streaming/Payloads/Disassemblers/PayloadDisassembler.cs:
```diff
@@ -120,12 +122,12 @@ private Task SendAsync()
         }
 
-        private async Task OnSent(Header header)
+        private async Task OnSentAsync(Header header)
```
>OnSentAsync [](start = 27, length = 11)

cancelationToken

gabog replied:
Can't add the paramter without breaking the interface, we will need to refactor this as a separate effort if we want to support cancellationTokens.

tomlm commented on libraries/Microsoft.Bot.Streaming/Payloads/Disassemblers/CancelDisassembler.cs:
```diff
@@ -22,7 +22,9 @@ public CancelDisassembler(IPayloadSender sender, Guid id, char type)
 
+#pragma warning disable UseAsyncSuffix // Use Async suffix  (we can't change this without breaking binary compat)
         public Task Disassemble()
+#pragma warning restore UseAsyncSuffix // Use Async suffix
```
>UseAsyncSuffix [](start = 24, length = 14)

cancelationToken

gabog replied:
It is not needed by any downstream call, we can add an optional parameter in the feuture when is required.

tomlm commented on libraries/Microsoft.Bot.Streaming/Payloads/Assemblers/ReceiveResponseAssembler.cs:
```diff
@@ -73,15 +73,15 @@ public void OnReceive(Header header, Stream stream, int contentLength)
         public void Close() => _streamManager.CloseStream(Id);
 
-        private async Task ProcessResponse(Stream stream)
+        private async Task ProcessResponseAsync(Stream stream)
```
>ProcessResponseAsync [](start = 27, length = 20)

cancelationToken

gabog replied:
Tried by couldn't. We don't have cancellationToken in the callers to this private method (adding it would break binary compat)

tomlm commented on libraries/Microsoft.Bot.Streaming/Payloads/Assemblers/ReceiveRequestAssembler.cs:
```diff
@@ -73,15 +73,15 @@ public void OnReceive(Header header, Stream stream, int contentLength)
         public void Close() => _streamManager.CloseStream(Id);
 
-        private async Task ProcessRequest(Stream stream)
+        private async Task ProcessRequestAsync(Stream stream)
```
>ProcessRequestAsync [](start = 27, length = 19)

cancelationToken

gabog replied:
The token is not available in any upstream callers and is not necessary in this private method.

tomlm commented:
>                     await _onCompleted(this.Id, request).ConfigureAwait(false);

cancelationToken

---
Refers to: libraries/Microsoft.Bot.Streaming/Payloads/Assemblers/ReceiveRequestAssembler.cs:120 in 625103c. [](commit_id = 625103c45f990b4c8d3a3d281dd47b5ae51461f2, deletion_comment = False)

tomlm commented on libraries/Microsoft.Bot.Streaming/Payloads/PayloadStream.cs:
```diff
@@ -131,9 +127,9 @@ public override int Read(byte[] buffer, int offset, int count)
             if (_active == null)
             {
-                dataAvailable.Wait();
+                _dataAvailable.Wait();
```
> _dataAvailable.Wait(); [](start = 15, length = 23)

eeesh, can't we throw an NotSupported exception on this?

DDEfromOR replied:
If the binary compatibility check will let us get away with it, NotSupported would make more sense. I think this was originally added when the class was still internal and we thought it may be useful to have. Now it's just a bear trap sitting in the middle of the hallway. At the least we could add a warning to the summary.

gabog commented:
I addressed some of Tom's comments and also updated the branch (and fixed a lot of conflicts).
I don't know enough about the library design to address the remaining comments from Tom (but I think we can leave it as is).
@DDEfromOR, if you can take a look today it would be great, it is getting painful to keep this branch fresh and resolve conflicts from master.

DDEfromOR commented on libraries/Microsoft.Bot.Streaming/Transport/NamedPipe/NamedPipeClient.cs:
```diff
@@ -109,8 +112,14 @@ public async Task ConnectAsync(IDictionary<string, string> requestHeaders)
 
+#pragma warning disable CA2000 // Dispose objects before losing scope
+
+            // We don't dispose the websocket, since NamedPipeTransport is now
```
Nit: there is no websocket involved here.
